primitives/externalities/src/lib.rs

```diff
@@ -228,6 +228,21 @@ pub trait Externalities: ExtensionStore {
 	/// no transaction is open that can be closed.
 	fn storage_commit_transaction(&mut self) -> Result<(), ()>;
 
+	/// Index specified transaction slice and store it.
+	fn storage_index_transaction(&mut self, _index: u32, _offset: u32) -> Result<(), ()> {
+		unimplemented!("storage_index_transaction");
+	}
+
+	/// Renew existing piece of transaction storage.
+	fn storage_renew_transaction_index(
```

Comment:
Is this really something that needs to be part of Substrate? We can not extend the externalities for every possible kind of blockchain that could exists out there.

Why can this not be done as part of the block import? By having some kind of special transaction that the client also recognizes? 

Or you could use an extension and we extend the extensions in such a way that they can contribute to the `StorageChanges`

Reply:
> Why can this not be done as part of the block import? By having some kind of special transaction that the client also recognizes?

Same reason we define all transactions in the runtime. So that transaction format could be upgraded. It can't be a completely special transaction as it must still be validated by the runtime. I.e. signature check, fees.

I've considered doing some kind of wrapper around `OpaqueExtrinsic` that would contain a part of the transaction, that the client could work with, along with the opaque part, but it turned out to be quite complicated because `construct_runtime` machinery expects a certain extrinsic structure. It also takes away some flexibility from the extrinsic format definition. So after discussing it with @gavofyork we've decided to go with externalities.

I'll look into moving it to extension though.

Reply:
> but it turned out to be quite complicated because `construct_runtime` machinery expects a certain extrinsic structure

This was a huge mistake by me. We should probably try to tackle this at some point to make this more flexible. As I will probably require some special extrinsic format for parachains in the near future, I will probably run into the same problems.

Reply:
Unfortunately using an extension is not really applicable here, as the collected changes need to be part of the execution context do be committed to the database along with the reset of the block data. So it needs to go into `OverlayedChanges`.  Offchain storage changes are in the externalities for the same reason I guess.

> Is this really something that needs to be part of Substrate? We can not extend the externalities for every possible kind of blockchain that could exists out there.

As I understand it - yes. The goal is for substrate to support such chains.